src/background/StreamSessionManager.ts

```diff
@@ -0,0 +1,373 @@
+        this.sessionState = {
+            leaderId: this.currentUserId!,
+            sessionJoinOrder: [this.currentUserId!],
+            sessionUsers: this.sessionState.sessionUsers
+        };
+        console.warn(this.sessionState);
+    }
+
+    /**
+     * Connects to a a peer. Can be used to start a session join if the current ConnectionState is Disconnected
+     * @param peerId The id of the peer to initially connect to
+     * @param [isReferral=false] Optional param that indicates that the peer we're connecting to is a referral
+     */
+    public connectToPeer(peerId: string, isReferral = false): void {
+        console.warn(`connect to peer: ${peerId} isReferral=${isReferral}`);
+        if (isReferral) {
+            this.connectionState = ConnectionState.ConnectingToPeer;
+            this.connectToExistingSessionReferralId = peerId;
+        }
+
+        const connectionMetadata: IConnectToPeerMetadata = {
+            // Todo - pass user display name (stored in localstorage?)
+            userDisplayName: "",
+            connectionState: this.connectionState
+        };
+        const conn = this.peer.connect(peerId, {
+            metadata: connectionMetadata
+        });
+
+        conn.on("open", () => this.handleConnect(conn, ""));
+    }
+
+    /**
+     * Releases managed resources
+     */
+    public destroy(): void {
+        this.peer.destroy();
+    }
+
+    /**
+     * Handle a new peer connection
+     * @param conn The underlying data connection
+     * @param userDisplayName The name of the user who initiated the connection
+     */
+    private handleConnect(conn: Peer.DataConnection, userDisplayName: string): void {
+        console.warn("handleConnect");
+        // Add new connection to local users
+        const user: IUserConnection = {
+            id: conn.peer,
+            displayName: userDisplayName,
+            connection: conn,
+            permissions: {
+                canControlVideo: true
+            }
+        };
+        this.sessionState.sessionUsers.set(user.id, user);
+
+        conn.on("data", (data: Message) => {
+            console.warn(`data from: ${conn.peer}`);
+            console.warn(data);
+            switch (data.messageType) {
+                case MessageType.StateSync:
+                    this.handleStateSyncMessage(data);
+                    break;
+                case MessageType.StateSyncRequest:
+                    this.handleStateSyncRequestMessage();
+                    break;
+                default:
+                    break;
+            }
+
+            this.port.postMessage(data);
+        });
+
+        conn.on("close", () => {
+            this.sessionState.sessionUsers.delete(user.id);
+            this.sessionState.sessionJoinOrder.splice(this.sessionState.sessionJoinOrder.indexOf(user.id), 1);
+            this.notifyUserCountChanges(-1, false);
+            if (user.id === this.sessionState.leaderId) {
```

Comment:
Feels like this might be best done abstracted into another `handleDisconect()` function to match other spots.